int64 width
int64 height
int64 end_x
int64 end_y
int64 current_x
int64 current_y
int64 current_value
